src/p11_wrap.cpp

```diff
@@ -0,0 +1,276 @@
+}
+
+/***********************************************************************************
+ * Implementation of P11_Session
+ */
+
+P11_Session::P11_Session(P11_PKCS11_SLOT_PTR slot, const std::string &pin, SessionMode mode)
+        : _slot(slot)
+{
+    openSession(mode);
+    login(pin);
+}
+
+P11_Session::~P11_Session() { logout(); }
+
+P11_PKCS11_TOKEN_PTR P11_Session::getTokenFromSlot()
+{
+    // Simply obtain the token via the slot's data structure.
+    if (_slot.get()->token == nullptr) {
+        throw P11Exception(P11Exception::nullTokenExceptionString);
+    }
+
+    // Leverage aliasing.
+    return P11_PKCS11_TOKEN_PTR(_slot, _slot.get()->token);
+}
+
+void P11_Session::openSession(SessionMode mode)
+{
+    P11IsNonNegative::callChecked(
+            lib::LibP11::P11_PKCS11_open_session, _slot.get(), mode /* Denotes read/write mode. */);
+}
+
+void P11_Session::login(const std::string &pin)
+{
+    if (isLoggedIn()) {
+        // Weird: We are trying to log in, when we are logged in already.
+        throw P11Exception(P11Exception::mismatchLoginExceptionString);
+    }
+
+    P11IsNonNegative::callChecked(lib::LibP11::P11_PKCS11_login,
+                                  _slot.get(),
+                                  0 /* Always login as normal user. */,
+                                  pin.c_str());
+}
+
+void P11_Session::logout()
+{
+    if (!isLoggedIn()) {
+        // Weird: We are trying to log out, when we are logged out already.
+        throw P11Exception(P11Exception::mismatchLoginExceptionString);
+    }
+
+    P11IsNonNegative::callChecked(lib::LibP11::P11_PKCS11_logout, _slot.get());
+}
+
+bool P11_Session::isLoggedIn()
+{
+    int result;
+    P11IsNonNegative::callChecked(
+            lib::LibP11::P11_PKCS11_is_logged_in, _slot.get(), 0 /* Check normal user. */, &result);
+
+    // Result is 1 if logged in.
+    return result == 1;
+}
+
+/***********************************************************************************
+ * Implementation of API functions.
+ */
+
+/* Destroys a PKCS11 context by first unloading it, and then freeing it. */
+static void _PKCS11_CTX_destroy(PKCS11_CTX *ctx)
+{
+    if (ctx == nullptr) {
+        throw P11Exception(P11Exception::nullCtxExceptionString);
+    }
+    lib::LibP11::P11_PKCS11_CTX_unload(ctx);
+    lib::LibP11::P11_PKCS11_CTX_free(ctx);
+}
+
+P11_PKCS11_CTX_Ptr _PKCS11_CTX_create(const std::string &module)
+{
+    PKCS11_CTX *ctx = P11CallPtr::callChecked(lib::LibP11::P11_PKCS11_CTX_new);
+    try {
+        P11IsNonNegative::callChecked(lib::LibP11::P11_PKCS11_CTX_load, ctx, module.c_str());
+    } catch (const P11Exception &e) {
+        // Free the ctx if loading failed and then throw.
+        lib::LibP11::P11_PKCS11_CTX_free(ctx);
+        throw;
+    }
+
+    // Create shared pointer which will unload and free the ctx automatically.
+    return P11_PKCS11_CTX_Ptr{ctx, _PKCS11_CTX_destroy};
+}
+
+static std::string parseID(const std::string &idHexString)
```

Comment:
This is just a simple helper function to convert IDs written in hex string by the user, and converts it to data so that it can be passed to libp11. The comment below explains this quite clearly imo.

As for:

> add some generic unhex utility function somehwhere

Not sure what the goal of this is. The code already makes use of a " generic unhex utility function", provided by Boost API.